modules/kafka/src/main/java/org/testcontainers/containers/KafkaContainer.java

```diff
@@ -80,83 +70,59 @@ public KafkaContainer withExternalZookeeper(String connectString) {
             command += "echo 'dataLogDir=/var/lib/zookeeper/log' >> zookeeper.properties\n";
             command += "zookeeper-server-start zookeeper.properties &\n";
         }
 
-        command += "export KAFKA_ZOOKEEPER_CONNECT='" + zookeeperConnect + "'\n";
-
-        command += "export KAFKA_ADVERTISED_LISTENERS='" + String.join(",", getBootstrapServers(), brokerAdvertisedListener(containerInfo)) + "'\n";
-
-        command += ". /etc/confluent/docker/bash-config \n";
-        command += "/etc/confluent/docker/configure \n";
-        command += "/etc/confluent/docker/launch \n";
+        // Optimization: skip the checks
+        command += "echo '' > /etc/confluent/docker/ensure \n";
+        // Run the original command
+        command += "/etc/confluent/docker/run \n";
+        withCommand("sh", "-c", command);
+    }
 
-        copyFileToContainer(
-            Transferable.of(command.getBytes(StandardCharsets.UTF_8), 0777),
-            STARTER_SCRIPT
+    @Override
+    @SneakyThrows
+    protected void containerIsStarted(InspectContainerResponse containerInfo) {
+        String brokerAdvertisedListener = brokerAdvertisedListener(containerInfo);
+        ExecResult result = execInContainer(
+            "kafka-configs",
+            "--alter",
+            "--bootstrap-server", brokerAdvertisedListener,
+            "--entity-type", "brokers",
+            "--entity-name", getEnvMap().get("KAFKA_BROKER_ID"),
+            "--add-config",
+            "advertised.listeners=[" + String.join(",", getBootstrapServers(), brokerAdvertisedListener) + "]"
         );
+        if (result.getExitCode() != 0) {
+            throw new IllegalStateException(result.getStderr());
+        }
     }
 
     protected String brokerAdvertisedListener(InspectContainerResponse containerInfo) {
-        // Kafka supports only one INTER_BROKER listener, so we have to pick one.
-        // The current algorithm uses the following order of resolving the IP:
-        // 1. Custom network's IP set via `withNetwork`
-        // 2. Bridge network's IP
-        // 3. Best effort fallback to getNetworkSettings#ipAddress
-        String ipAddress = containerInfo.getNetworkSettings().getNetworks().entrySet()
-            .stream()
-            .filter(it -> it.getValue().getIpAddress() != null)
-            .max(Comparator.comparingInt(entry -> {
-                if (getNetwork() != null && getNetwork().getId().equals(entry.getValue().getNetworkID())) {
-                    return 2;
-                }
-
-                if ("bridge".equals(entry.getKey())) {
-                    return 1;
-                }
-
-                return 0;
-            }))
-            .map(it -> it.getValue().getIpAddress())
-            .orElseGet(() -> containerInfo.getNetworkSettings().getIpAddress());
-
-        return String.format("BROKER://%s:%s", ipAddress, "9092");
+        return String.format("BROKER://%s:%s", containerInfo.getConfig().getHostName(), "9092");
```

Comment:
this was super fun to discover, but apparently we over-engineered this, and `hostName` is enough to make it work reliably - it works with both "no network" (a.k.a. "bridge only") and multi network scenarios, as containers are visible by their hostname to other containers in the same network 😂